add a DataImport class that will pick implementation based on source URL

grishick opened the pull request:
@desouzas pointed out that passing different DataImport implementation classname to ZMailbox::createDataSource based on the source is unnecessary. Each contacts API is unique and therefore the mapping cannot change without changing the code, so we may as well store the mapping inside the code.
With this change the oauth2 service will always call ZMailbox::createDataSource will "com.synacor.zimbra.OAuthDataImport" as the value for "importClass" - as in the SOAP test in this extension.

tmclane commented:
Only suggestion I have is that perhaps the URL <-> Implementation could be defined in a configuration file so that a URL change doesn't require a recompile?

grishick commented:
@tmclane fair enough :) I thought I could get away without it :)

desouzas commented on src/java/com/synacor/zimbra/OAuthDataImport.java:
```diff
@@ -0,0 +1,35 @@
+    private DataImport implementation;
+    public OAuthDataImport(DataSource ds) {
+        String sourceURL = ds.getOauthRefreshTokenUrl();
+        if(sourceURL.indexOf("api.login.yahoo.com/oauth2/get_token") > -1) {
```
Let's find another way to determine which service we need. Something more generic so it's easier to reuse this stuff for syncs beyond contacts (e.g. calendar, profile, photos, ...).

Also let's talk about minimizing storage of the refreshTokenUrl since it doesn't change by user.

grishick commented:
@desouzas I've updated this pull request with changes that get rid of refreshToken URL parameter and add host constant for yahoo. I also renamed LC settings to match what you are using. Additionally, I got rid of oauth-specific classes and constants, because all the features fit into generic datasource implementation.
There is an example of creating a datasource for importing Yahoo contacts in TestYahooContactsImport.java:

```
        ZMailbox zmbox = TestUtil.getZMailbox(USER_NAME);
        ZFolder folder = TestUtil.createFolder(zmbox, "/testCustomDS");
        ZDataSource zds = new ZDataSource(yahooTestAccount, true);
        zds.setImportClass("com.synacor.zimbra.OAuthDataImport");
        zds.setRefreshToken(testToken);
        zds.setHost(ZDataSource.SOURCE_HOST_YAHOO);
        zds.setFolderId(folder.getId());
        String dsId = zmbox.createDataSource(zds);
```

once I finish implementing importData method, there will be another example that triggers one time import. After that the last step will be to implement scheduling using ZCS internal mechanism, so you don't have to use cron for that.

tmclane commented:
One question: Do we care if the 'author' is added to the files in the comments

tmclane commented:
Oops didn't mean to close it.
